crates/ntx-builder/src/block_producer.rs

```diff
@@ -1,44 +1,51 @@
-use std::net::SocketAddr;
 use std::time::Duration;
 
 use futures::{TryStream, TryStreamExt};
+use miden_node_proto::clients::{
+    BlockProducer,
+    BlockProducerClient as InnerBlockProducerClient,
+    Builder,
+};
 use miden_node_proto::domain::mempool::MempoolEvent;
-use miden_node_proto::generated::block_producer::api_client::ApiClient;
 use miden_node_proto::generated::{self as proto};
 use miden_node_utils::FlattenResult;
-use miden_node_utils::tracing::grpc::OtelInterceptor;
 use miden_objects::block::BlockNumber;
 use miden_objects::transaction::ProvenTransaction;
 use miden_tx::utils::Serializable;
 use tokio_stream::StreamExt;
 use tonic::Status;
-use tonic::service::interceptor::InterceptedService;
-use tonic::transport::Channel;
 use tracing::{info, instrument};
+use url::Url;
 
 use crate::COMPONENT;
 
-type InnerClient = ApiClient<InterceptedService<Channel, OtelInterceptor>>;
+// CLIENT
+// ================================================================================================
 
 /// Interface to the block producer's gRPC API.
 ///
 /// Essentially just a thin wrapper around the generated gRPC client which improves type safety.
 #[derive(Clone, Debug)]
 pub struct BlockProducerClient {
-    inner: InnerClient,
+    client: InnerBlockProducerClient,
 }
 
 impl BlockProducerClient {
     /// Creates a new block producer client with a lazy connection.
-    pub fn new(block_producer_address: SocketAddr) -> Self {
-        // SAFETY: The block_producer_url is always valid as it is created from a `SocketAddr`.
-        let block_producer_url = format!("http://{block_producer_address}");
-        let channel =
-            tonic::transport::Endpoint::try_from(block_producer_url).unwrap().connect_lazy();
-        let block_producer = ApiClient::with_interceptor(channel, OtelInterceptor);
-        info!(target: COMPONENT, block_producer_endpoint = %block_producer_address, "Store client initialized");
-
-        Self { inner: block_producer }
+    pub fn new(block_producer_url: &Url) -> Self {
+        // SAFETY: The block_producer_url is always valid as it is a user-provided URL that has been
+        // validated.
+        let block_producer = Builder::new(block_producer_url.to_string())
+            .expect("Failed to initialize block-producer endpoint")
+            .without_tls()
+            .without_timeout()
+            .without_metadata_version()
+            .without_metadata_genesis()
+            .connect_lazy::<BlockProducer>();
+
+        info!(target: COMPONENT, block_producer_endpoint = %block_producer_url, "Block producer client initialized");
```

Comment:
nit: I think these were like this before, but I wonder how much value these add. The client was created but it connects lazily, right? So it would effectively always initialize correctly. Maybe not something to change, but I think it could look misleading in logs unless you always keep in mind that these are lazy clients.

Reply:
I've changed the message to better reflect this. What do you think?